marshmallow/fields.py

```diff
@@ -1139,11 +1147,91 @@ class Dict(Field):
         'invalid': 'Not a valid mapping type.'
     }
 
-    def _deserialize(self, value, attr, data):
-        if isinstance(value, collections.Mapping):
-            return value
+    def __init__(self, values=None, keys=None, **kwargs):
+        super(Dict, self).__init__(**kwargs)
+        if values is None:
+            self.value_container = None
+        elif isinstance(values, type):
+            if not issubclass(values, FieldABC):
+                raise ValueError('"values" must be a subclass of '
+                                 'marshmallow.base.FieldABC')
+            self.value_container = values()
+        else:
+            if not isinstance(values, FieldABC):
+                raise ValueError('"values" must be of type '
+                                 'marshmallow.base.FieldABC')
+            self.value_container = values
+        if keys is None:
+            self.key_container = None
+        elif isinstance(keys, type):
+            if not issubclass(keys, FieldABC):
+                raise ValueError('"keys" must be a subclass of '
+                                 'marshmallow.base.FieldABC')
+            self.key_container = keys()
         else:
+            if not isinstance(keys, FieldABC):
+                raise ValueError('"keys" must be of type '
+                                 'marshmallow.base.FieldABC')
+            self.key_container = keys
+
+    def _serialize(self, value, attr, obj):
+        if value is None:
+            return None
+        if not self.value_container and not self.key_container:
+            return value
+        if isinstance(value, collections.Mapping):
+            items = value.values()
+            if self.value_container:
+                items = [self.value_container._serialize(item, attr, obj) for item in items]
+            keys = value.keys()
+            if self.key_container:
+                keys = [self.key_container._serialize(key, attr, obj) for key in keys]
+            return dict(zip(keys, items))
+        self.fail('invalid')
+
+    def _deserialize(self, value, attr, data):
+        if not isinstance(value, collections.Mapping):
             self.fail('invalid')
+        if not self.value_container and not self.key_container:
+            return value
+
+        result = {}
+        errors = {}
+        items = list(value.values())
+        keys = list(value.keys())
+        if self.key_container:
+            for idx, key in enumerate(keys):
+                try:
+                    keys[idx] = self.key_container.deserialize(key)
+                except ValidationError as e:
+                    key_messages = [
+                        'Invalid key: {}'.format(message)
+                        for message in e.messages
+                    ]
+                    result.update({key: list(key_messages)})
```

Comment:
Also, shouldn't the `result` dict take `e.data`?

Reply:
In fact, nothing should be written to `result` - that is overwritten at the end anyway by `result = dict(zip(keys, items))`. Shouldn't this be `keys[idx] = e.data`? Otherwise that zip will fail (or worse - with one bad key and one bad value independently, the zip will match up incorrect keys to values)